lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -224,65 +233,61 @@ export default class BitcoinProcessor {
-        BitcoinProcessor.addBlockToChainInfo(chainInfos, indexInChainInfos, block);
+        const blockReward = BitcoinProcessor.getBitcoinBlockReward(block.height);
+        notYetValidatedBlocks.set(
+          block.hash,
+          {
+            height: block.height,
+            hash: block.hash,
+            totalFee: block.transactions[0].outputs[0].satoshis - blockReward,
+            transactionCount: block.transactions.length - 1, // minus one because of coinbase
+            previousHash: block.previousHash }
+          );
         await this.processSidetreeTransactionsInBlock(block);
       }
     }
   }
 
   /**
-   * Checks in the chain info to find the earliest valid block and its hash,
-   * and delete all transactions from invalid blocks in DB.
+   * Find all hashes in the notYetValidatedBlocks that are actually valid,
+   * add them to the validated list and delete them from the map.
    */
-  private async findEarliestValidBlock (
-    chainInfos: Map<string, IBlockInfoExtended>[],
+  private async findEarliestValidBlockAndAddToValidBlocks (
+    validatedBlocks: (IBlockInfoExtended | undefined)[],
+    notYetValidatedBlocks: Map<string, IBlockInfoExtended>,
     heightOfEarliestKnownValidBlock: number,
     hashOfEarliestKnownValidBlock: string,
     startingBlockHeight: number): Promise<[number, string]> {
 
     let index = heightOfEarliestKnownValidBlock - startingBlockHeight;
-    let currentFeeData = chainInfos[index];
-    while (currentFeeData !== undefined && currentFeeData.get(hashOfEarliestKnownValidBlock) !== undefined) {
+    let validBlock = notYetValidatedBlocks.get(hashOfEarliestKnownValidBlock);
+    while (validBlock !== undefined && index >= 0) {
       console.log(`Found valid block ${hashOfEarliestKnownValidBlock}`);
-      // delete all unneeded key value pairs and db records
-      const hashes = currentFeeData.keys();
-      for (let hash of hashes) {
-        if (hash !== hashOfEarliestKnownValidBlock) {
-          console.log(`Deleting data for block ${hash} because it has the same height as ${hashOfEarliestKnownValidBlock}`);
-          currentFeeData.delete(hash);
-          await this.transactionStore.removeTransactionByTransactionTimeHash(hash);
-        }
-      }
-      const confirmedFeeData = currentFeeData.get(hashOfEarliestKnownValidBlock)!;
-      hashOfEarliestKnownValidBlock = confirmedFeeData.previousHash;
+      validatedBlocks[index] = validBlock;
+      // delete because it is now validated
+      notYetValidatedBlocks.delete(hashOfEarliestKnownValidBlock);
+      // the previous block hash becomes valid
+      hashOfEarliestKnownValidBlock = validBlock.previousHash;
       heightOfEarliestKnownValidBlock--;
       index--;
-      currentFeeData = chainInfos[index];
+      validBlock = notYetValidatedBlocks.get(hashOfEarliestKnownValidBlock);
     }
     return [heightOfEarliestKnownValidBlock, hashOfEarliestKnownValidBlock];
```

Comment:
Comment: it just pains me that we are forced to pass in AND return this tuple because of the initialization case... 'coz except for the first block, `validatedBlocks` should already give us info needed.... I have no good solution to this, but thought I should call out in case I missed something obvious.

Reply:
Changed